sw/device/silicon_creator/lib/shutdown.c

```diff
@@ -445,30 +446,46 @@ SHUTDOWN_FUNC(noreturn, shutdown_hang(void)) {
 #ifdef OT_PLATFORM_RV32
   while (true) {
     asm volatile(
-        "1:"
+        ".L_shutdown_hang_asm_start:"
         // Request a new scrambling key.
         "sw %[kSramRenewKey], %[kSramCtrlCtrlReg](%[kSramCtrlBase]);"
         // Request a system reset.
         "sw %[kMultiBitBool4True], %[kRstmgrResetReqReg](%[kRstmgrBase]);"
 
-        // TODO(lowRISC/opentitan#7148): restrict the ePMP such that only
-        // ROM may execute.  mundaym's suggestion: set entry 2 as a NAPOT
-        // region covering the entire address space, clear all its permission
-        // bits and set the lock bit, and then finally disable RLB to prevent
-        // any further modifications.
+        // Reconfigure ePMP such that only this inline asm is executable:
+        // - Entry 0: TOR, LRX, [.L_shutdown_hang_asm_start,
+        // .L_shutdown_hang_asm_end).
+        // - Entry 1: NAPOT, L, entire address space.
+        // - MSECCFG: MMWP set RLB unset to prevent any further modifications.
+        "la   t0, .L_shutdown_hang_asm_start;"
+        "csrw pmpaddr0, t0;"
+        "la   t0, .L_shutdown_hang_asm_end;"
+        "csrw pmpaddr1, t0;"
+        "csrw pmpaddr2, %[kNapotEntireAddressSpace];"
+        "csrw pmpcfg0, %[kPmpCfg0];"
+        "csrw pmpcfg1, zero;"
+        "csrw pmpcfg2, zero;"
+        "csrw pmpcfg3, zero;"
+        "csrw %[kMSecCfgReg], %[kMSecCfgVal];"
 
         // Generate a halt-maze.
-        "wfi; wfi; wfi; wfi; j 1b;"
-        "wfi; wfi; j 1b;"
-        "wfi; j 1b;"
+        "wfi; wfi; wfi; wfi; j .L_shutdown_hang_asm_start;"
+        "wfi; wfi; j .L_shutdown_hang_asm_start;"
+        "wfi; j .L_shutdown_hang_asm_start;"
         "wfi;"
+        ".L_shutdown_hang_asm_end:"
         :
         : [kSramRenewKey] "r"(1 << SRAM_CTRL_CTRL_RENEW_SCR_KEY_BIT),
           [kSramCtrlCtrlReg] "I"(SRAM_CTRL_CTRL_REG_OFFSET),
           [kSramCtrlBase] "r"(kSramCtrlBase),
           [kMultiBitBool4True] "r"(kMultiBitBool4True),
+          [kRstmgrResetReqReg] "I"(RSTMGR_RESET_REQ_REG_OFFSET),
           [kRstmgrBase] "r"(kRstmgrBase),
-          [kRstmgrResetReqReg] "I"(RSTMGR_RESET_REQ_REG_OFFSET));
+          [kNapotEntireAddressSpace] "r"(0x7fffffff),
+          [kPmpCfg0] "r"((EPMP_CFG_A_TOR | EPMP_CFG_LRX) |
```

Comment:
If you are using pmpaddr0 and pmpaddr1 for a TOR region then you want the config in pmp1cfg (i.e. shifted left by 8). Right now this is configuring pmp0cfg which will use the implicit starting address 0 as the bottom-of-range and pmpaddr0 as the top-of-range.

Reply:
Done, thanks!